mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,145 +166,301 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
 
-                Value sizeArray =
-                    rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
+            unpackMemRef(result, qJacobian, callArgs, rewriter, loc, {.dupNoNeed = true});
+        }
 
-                vectorIndices[0] = 4;
-                Value strideArray =
-                    rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
+        // The results of backprop are in data in
+        rewriter.create<LLVM::CallOp>(loc, backpropFnDecl, callArgs);
+        rewriter.eraseOp(op);
+        return success();
+    }
 
-                bufferSize = rewriter.create<LLVM::MulOp>(loc, sizeArray, strideArray);
-                bufferSize = rewriter.create<LLVM::AddOp>(loc, offset, bufferSize);
-            }
+  private:
+    static FlatSymbolRefAttr getOrInsertEnzymeGlobal(OpBuilder &builder, ModuleOp moduleOp,
+                                                     const char *globalName)
+    {
+        // Copyright (C) 2023 - Jacob Mai Peng
+        // https://github.com/pengmai/lagrad/blob/main/lib/LAGrad/LowerToLLVM.cpp
+        auto *context = moduleOp.getContext();
+        if (moduleOp.lookupSymbol<LLVM::GlobalOp>(globalName)) {
+            return SymbolRefAttr::get(context, globalName);
+        }
 
-            Value bufferMemSize = rewriter.create<LLVM::MulOp>(loc, bufferSize, memrefArgSize);
+        OpBuilder::InsertionGuard insertGuard(builder);
+        builder.setInsertionPointToStart(moduleOp.getBody());
+        auto shortTy = IntegerType::get(context, 8);
+        builder.create<LLVM::GlobalOp>(moduleOp.getLoc(), shortTy,
+                                       /*isConstant=*/true, LLVM::Linkage::Linkonce, globalName,
+                                       IntegerAttr::get(shortTy, 0));
+        return SymbolRefAttr::get(context, globalName);
+    }
 
-            Value buffer =
-                rewriter.create<LLVM::CallOp>(loc, allocFnDecl, bufferMemSize).getResult();
-            // Set value to 0 (gradients)
-            Value c0 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(0));
-            // Value c0float = rewriter.create<LLVM::ConstantOp>(loc,
-            // rewriter.getF64FloatAttr(0.0));
-            rewriter.create<LLVM::CallOp>(loc, memsetFnDecl,
-                                          ArrayRef<Value>{buffer, c0, bufferMemSize});
+    Value castToConvertedType(Value value, OpBuilder &builder, Location loc) const
+    {
+        auto casted = builder.create<UnrealizedConversionCastOp>(
+            loc, getTypeConverter()->convertType(value.getType()), value);
+        return casted.getResult(0);
+    }
 
-            auto llvmInsert0 = rewriter.create<LLVM::InsertValueOp>(loc, llvmDataIn, buffer, 0);
-            auto llvmInsert1 = rewriter.create<LLVM::InsertValueOp>(loc, llvmInsert0, buffer, 1);
+    void unpackMemRef(Value memRefArg, Value shadowMemRef, SmallVectorImpl<Value> &callArgs,
```

Comment:
Minor: would be nice if the name of the function indicated that it is modifying one of the args since it doesn't return anything, like `unpackMemRefAndAppend`.

Reply:
Done